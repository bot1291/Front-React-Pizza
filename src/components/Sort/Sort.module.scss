@import '../../styles/variables.scss';

.sort {
	position: relative;

	.label {
		display: flex;
		align-items: center;

		svg {
			margin-right: 8px;
		}

		b {
			margin-right: 8px;
		}

		span {
			color: $orange;
			border-bottom: 1px dashed $orange;
			cursor: pointer;
		}
	}

	.popup {
		position: absolute;
		right: 0;
		width: 160px;
		margin-top: 15px;
		padding: 10px 0;
		overflow: hidden;
		background: $white;
		border-radius: 10px;
		box-shadow: 0 5px 15px rgba(0, 0, 0, 0.1%);

		ul {
			overflow: hidden;

			li {
				padding: 12px 20px;
				cursor: pointer;

				&:hover {
					background: rgba($color: $black, $alpha: 0.05%);
				}

				&.active {
					color: $orange;
					font-weight: bold;
					background: rgba($color: $black, $alpha: 0.05%);
				}
			}
		}
	}
}
