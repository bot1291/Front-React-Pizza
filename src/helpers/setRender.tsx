import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createReduxStore } from '../store/store';

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const setRender = <T,>(Component: JSX.Element, store?: Record<string, T>) => {
	return render(
		<Provider store={createReduxStore(store)}>{Component}</Provider>
	);
};

export default setRender;
