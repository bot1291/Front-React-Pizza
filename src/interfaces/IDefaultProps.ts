import { TypeParams } from './TypeParams';

export interface IDefaultProps {
	sortParams: TypeParams[];
}
