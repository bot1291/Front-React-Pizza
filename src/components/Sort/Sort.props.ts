import { DetailedHTMLProps, HTMLAttributes } from 'react';
import { TypeParams } from '../../interfaces/TypeParams';

export interface SortProps
	extends DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement> {
	sortParams: TypeParams[];
}
