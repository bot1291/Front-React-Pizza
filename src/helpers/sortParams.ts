import { TypeParams } from '../interfaces/TypeParams';

const sortParams: TypeParams[] = [
	{ param: 'rating', title: 'популярности' },
	{ param: 'currentPrice', title: 'цене' },
	{ param: 'title', title: 'алфавиту' },
];

export default sortParams;
