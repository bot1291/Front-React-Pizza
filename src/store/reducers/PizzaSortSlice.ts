import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { TypeParams } from '../../interfaces/TypeParams';
import { IPizza } from '../../interfaces/IPizza';
import { fetchPizzas } from './data';

const allPizzaTypes = fetchPizzas.map((p) => p.types).flat();
const sortedTypes: string[] = [];
allPizzaTypes.forEach((p) => {
	if (!sortedTypes.includes(p)) {
		sortedTypes.push(p);
	}
});

interface IUserState {
	isLoading: boolean;
	error: string;
	pizzas: IPizza[];
	pizzasBackup: IPizza[];
	allPizzaTypes: string[];
	currentType: string;
	searchValue: string;
	currentSortParam: TypeParams;
	isSortedByType: boolean;
	isSortedToDown: boolean;
}

export const initialState: IUserState = {
	isLoading: false,
	error: '',
	pizzas: localStorage.pizzas
		? (JSON.parse(localStorage.pizzas) as unknown as IPizza[])
		: fetchPizzas,
	pizzasBackup: localStorage.pizzasBackup
		? JSON.parse(localStorage.pizzasBackup)
		: fetchPizzas,
	allPizzaTypes: sortedTypes,
	currentType: 'все',
	searchValue: '',
	isSortedByType: false,
	isSortedToDown: true,
	currentSortParam: {
		title: 'популярности',
		param: 'rating',
	},
};

export type PossibleSortParams = 'price' | 'title' | 'rating';

const checkForSimilarity = (word: string, search: string): boolean => {
	return word
		.toLowerCase()
		.split(' ')
		.join('')
		.includes(search.toLowerCase().split(' ').join(''));
};

export const pizzaSortSlice = createSlice({
	name: 'pizza',
	initialState,
	reducers: {
		setParam: (state, action: PayloadAction<IPizza[]>) => {
			state.pizzas = action.payload;
			state.pizzasBackup = state.pizzasBackup.map((pizzaBack) => {
				const replacementPizza = action.payload.find(
					(pizza) => pizzaBack.title === pizza.title
				);
				if (replacementPizza) {
					return replacementPizza;
				}
				return pizzaBack;
			});
		},
		sortToUpOrDown: (state, action: PayloadAction<boolean>) => {
			state.pizzas = state.pizzas.sort((a, b) => {
				if (
					a[state.currentSortParam.param] >
					b[state.currentSortParam.param]
				) {
					return action.payload ? -1 : 1;
				}
				return action.payload ? 1 : -1;
			});
			state.isSortedToDown = !state.isSortedToDown;
		},
		sortByParam: (state, action: PayloadAction<TypeParams>) => {
			state.pizzas = state.pizzas.sort((a, b) =>
				a[action.payload.param] > b[action.payload.param] ? 1 : -1
			);
			state.currentSortParam = action.payload;
			state.pizzasBackup = state.pizzasBackup.sort((a, b) =>
				a[action.payload.param] > b[action.payload.param] ? 1 : -1
			);
			state.isSortedToDown = true;
		},
		sortByType: (state, action: PayloadAction<string>) => {
			state.pizzas = state.pizzasBackup.filter((p) =>
				p.types.includes(action.payload)
			);
			state.currentType = action.payload;
			state.searchValue = '';
			state.isSortedByType = true;
		},
		unset: (state) => {
			state.pizzas = state.pizzasBackup;
			state.currentType = 'все';
			state.searchValue = '';
			state.isSortedByType = false;
		},
		search: (state, action: PayloadAction<string>) => {
			state.pizzas = (
				state.searchValue ? state.pizzasBackup : state.pizzas
			).filter(
				(p) =>
					checkForSimilarity(p.title, action.payload) ||
					checkForSimilarity(p.types.join(''), action.payload)
			);
		},
		setSearchValue: (state, action: PayloadAction<string>) => {
			state.searchValue = action.payload;
		},
	},
});

export default pizzaSortSlice.reducer;
