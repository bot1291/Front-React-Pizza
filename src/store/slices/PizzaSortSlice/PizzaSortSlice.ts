import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { TypeParams } from '../../../interfaces/TypeParams';
import { IPizza } from '../../../interfaces/IPizza';
import { fetchPizzas } from '../ActionCreators';
import checkForSimilarity from '../../../helpers/checkForSimilarity';

export interface IUserState {
	isLoading: boolean;
	error: boolean;
	pizzas: IPizza[];
	pizzasBackup: IPizza[];
	allPizzaTypes: string[];
	currentType: string;
	searchValue: string;
	currentSortParam: TypeParams;
	isSortedToDown: boolean;
}

export const initialState: IUserState = {
	isLoading: true,
	error: false,
	pizzas: [],
	pizzasBackup: [],
	allPizzaTypes: [],
	currentType: 'все',
	searchValue: '',
	isSortedToDown: localStorage.isDown === 'true' ? true : false,
	currentSortParam: {
		title: 'популярности',
		param: 'rating',
	},
};

export const pizzaSortSlice = createSlice({
	name: 'pizza',
	initialState,
	reducers: {
		setParam: (state, action: PayloadAction<IPizza[]>) => {
			state.pizzas = action.payload;
			state.pizzasBackup = state.pizzasBackup.map((pizzaBack) => {
				const replacementPizza = action.payload.find(
					(pizza) => pizzaBack.title === pizza.title
				);
				if (replacementPizza) {
					return replacementPizza;
				}
				return pizzaBack;
			});
		},
		sortToUpOrDown: (state, action: PayloadAction<boolean>) => {
			state.pizzas = state.pizzas.sort((a, b) => {
				if (
					a[state.currentSortParam.param] >
					b[state.currentSortParam.param]
				) {
					return action.payload ? -1 : 1;
				}
				return action.payload ? 1 : -1;
			});
			state.isSortedToDown = !action.payload;
		},
		sortByParam: (state, action: PayloadAction<TypeParams>) => {
			state.pizzas = state.pizzas.sort((a, b) =>
				a[action.payload.param] > b[action.payload.param] ? 1 : -1
			);
			state.pizzasBackup = state.pizzasBackup.sort((a, b) =>
				a[action.payload.param] > b[action.payload.param] ? 1 : -1
			);
			state.currentSortParam = action.payload;
		},
		sortByType: (state, action: PayloadAction<string>) => {
			state.pizzas = state.pizzasBackup.filter((p) =>
				p.types.includes(action.payload)
			);
			state.currentType = action.payload;
			state.searchValue = '';
		},
		unset: (state) => {
			state.pizzas = state.pizzasBackup;
			state.currentType = 'все';
			state.searchValue = '';
		},
		search: (state, action: PayloadAction<string>) => {
			state.pizzas = state.pizzasBackup.filter(
				(p) =>
					checkForSimilarity(p.title, action.payload) ||
					checkForSimilarity(p.types.join(''), action.payload)
			);
		},
		setSearchValue: (state, action: PayloadAction<string>) => {
			state.searchValue = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder.addCase(fetchPizzas.fulfilled, (state, action) => {
			state.isLoading = false;
			state.error = false;
			state.pizzas = action.payload?.pizzas ?? [];
			state.allPizzaTypes = action.payload?.types ?? [];
			state.pizzasBackup = action.payload?.pizzas ?? [];
		});
		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		builder.addCase(fetchPizzas.pending, (state, _action) => {
			state.isLoading = true;
		});
		// eslint-disable-next-line @typescript-eslint/no-unused-vars
		builder.addCase(fetchPizzas.rejected, (state, _action) => {
			state.isLoading = false;
			state.error = true;
		});
	},
});

export default pizzaSortSlice.reducer;

export const {
	search,
	setParam,
	setSearchValue,
	sortByParam,
	sortByType,
	sortToUpOrDown,
	unset,
} = pizzaSortSlice.actions;
