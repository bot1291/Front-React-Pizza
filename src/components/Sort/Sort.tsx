import styles from './Sort.module.scss';
import { SortProps } from './Sort.props';
import cn from 'classnames';
import { FC, useEffect, useState } from 'react';
import { useAppDispatch, useAppSelector } from '../../hooks/redux';
import { pizzaSortSlice } from '../../store/reducers/PizzaSortSlice';
import { TypeParams } from '../../interfaces/TypeParams';
import { IconContext } from 'react-icons';
import { AiFillCaretUp } from 'react-icons/ai';

export const Sort: FC<SortProps> = ({ sortParams, className, ...props }) => {
	const [isOpened, setIsOpened] = useState<boolean>(false);
	const dispatch = useAppDispatch();
	const { sortByParam, sortToUpOrDown } = pizzaSortSlice.actions;
	const { currentSortParam, isSortedToDown } = useAppSelector(
		(state) => state.pizzaSortReducer
	);

	useEffect(() => {
		if (localStorage.currentSortParam) {
			dispatch(sortByParam(JSON.parse(localStorage.currentSortParam)));
		} else {
			dispatch(sortByParam(currentSortParam));
		}
	}, [dispatch, sortByParam, sortToUpOrDown]);

	const chooseSort = (param: TypeParams) => {
		setIsOpened(false);
		localStorage.currentSortParam = JSON.stringify(param);
		dispatch(sortByParam(param));
	};
	const handlerSetIsSortedByUpOrDown = () => {
		dispatch(sortToUpOrDown(isSortedToDown));
	};
	return (
		<div className={cn(className, styles.sort)} {...props}>
			<button
				onClick={handlerSetIsSortedByUpOrDown}
				className={styles.icon}>
				<IconContext.Provider
					value={{
						size: '30px',
						className: cn({
							[styles.down]: isSortedToDown,
						}),
					}}>
					<AiFillCaretUp />
				</IconContext.Provider>
			</button>
			<button
				data-testid="sorting-button"
				onClick={() => setIsOpened(!isOpened)}
				className={cn(styles.label)}>
				<b>Сортировка по:</b>
				<span>{currentSortParam.title}</span>
			</button>
			<div
				data-testid="popup"
				className={cn(styles.popup, {
					[styles.popupActive]: isOpened,
				})}>
				<div className={styles.sortBlock}>
					{sortParams.map((p) => (
						<button
							key={p.param}
							onClick={() => chooseSort(p)}
							className={cn({
								[styles.active]:
									currentSortParam.param === p.param,
							})}>
							{p.title}
						</button>
					))}
				</div>
			</div>
		</div>
	);
};
