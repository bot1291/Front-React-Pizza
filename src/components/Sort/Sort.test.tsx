import { screen } from '@testing-library/react';
import { Sort } from './Sort';
import userEvent from '@testing-library/user-event';
import { initialState } from '../../store/slices/PizzaSortSlice/PizzaSortSlice';
import setRender from '../../helpers/setRender';
import { IDefaultProps } from '../../interfaces/IDefaultProps';
import sortParams from '../../helpers/sortParams';

const defaultProps: IDefaultProps = {
	sortParams: [
		{ param: 'rating', title: 'популярности' },
		{ param: 'currentPrice', title: 'цене' },
		{ param: 'title', title: 'алфавиту' },
	],
};

describe('Sort table where you can sort to up or down or by using params', () => {
	it('All pages / part pages performance', async () => {
		setRender(<Sort {...defaultProps} />, {
			pizzaSortReducer: {
				...initialState,
				isAllPages: false,
			},
		});
		const allPageButton = screen.getByTestId('all-pages-button');
		const partPageButton = screen.getByTestId('part-pages-button');

		expect(partPageButton).toHaveClass('active');

		await userEvent.click(allPageButton);
		expect(partPageButton).not.toHaveClass('active');
		expect(allPageButton).toHaveClass('active');
	});

	it('Other state for currentSortParam', async () => {
		setRender(<Sort {...defaultProps} />, {
			pizzaSortReducer: {
				...initialState,
				currentSortParam: {
					title: 'цене',
					param: 'currentPrice',
				},
			},
		});
		const buttonElem = screen.getByTestId('open-popuop-button');
		expect(buttonElem).toHaveTextContent('цене');
		const sortButtonElems = screen.getAllByTestId('sort-button');

		await userEvent.click(buttonElem);
		expect(screen.getByTestId('popup')).toHaveClass('popupActive');

		sortButtonElems[2] && (await userEvent.click(sortButtonElems[2]));
		expect(buttonElem).toHaveTextContent('алфавиту');

		sortButtonElems[0] && (await userEvent.click(sortButtonElems[0]));
		expect(buttonElem).toHaveTextContent('популярности');
	});

	it('Other state for isSortedToDown', async () => {
		setRender(<Sort {...defaultProps} />, {
			pizzaSortReducer: {
				...initialState,
				currentSortParam: {
					title: 'цене',
					param: 'currentPrice',
				},
				isSortedToDown: true,
			},
		});
		const sortUpDownButton = screen.getByTestId('sort-up-down-button');
		expect(sortUpDownButton).toHaveClass('down');

		await userEvent.click(sortUpDownButton);
		expect(sortUpDownButton).not.toHaveClass('down');
	});

	it('Sort up or down button', async () => {
		setRender(<Sort {...defaultProps} />);
		const sortUpDownButton = screen.getByTestId('sort-up-down-button');
		expect(sortUpDownButton).toHaveClass('down');

		await userEvent.click(sortUpDownButton);
		expect(sortUpDownButton).not.toHaveClass('down');

		await userEvent.click(sortUpDownButton);
		expect(sortUpDownButton).toHaveClass('down');
	});

	it('Popup opening', async () => {
		setRender(<Sort {...defaultProps} />);
		const buttonElem = screen.getByTestId('open-popuop-button');
		expect(buttonElem).toBeInTheDocument();
		expect(screen.getByTestId('popup')).not.toHaveClass('popupActive');

		await userEvent.click(buttonElem);
		expect(screen.getByTestId('popup')).toHaveClass('popupActive');

		await userEvent.click(buttonElem);
		expect(screen.getByTestId('popup')).not.toHaveClass('popupActive');
	});

	it('Sort buttons performance', async () => {
		setRender(<Sort sortParams={sortParams} />);
		const buttonElem = screen.getByTestId('open-popuop-button');
		expect(buttonElem).toHaveTextContent('популярности');
		const sortButtonElems = screen.getAllByTestId('sort-button');
		expect(sortButtonElems[0]).toBeInTheDocument();

		await userEvent.click(buttonElem);
		expect(screen.getByTestId('popup')).toHaveClass('popupActive');

		sortButtonElems[2] && (await userEvent.click(sortButtonElems[2]));
		expect(buttonElem).toHaveTextContent('алфавиту');

		sortButtonElems[0] && (await userEvent.click(sortButtonElems[0]));
		expect(buttonElem).toHaveTextContent('популярности');

		sortButtonElems[1] && (await userEvent.click(sortButtonElems[1]));
		expect(buttonElem).toHaveTextContent('цене');
		expect(sortButtonElems).toMatchSnapshot();
	});
});
