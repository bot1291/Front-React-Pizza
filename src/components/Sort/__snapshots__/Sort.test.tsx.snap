// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort table where you can sort to up or down or by using params Sort buttons performance 1`] = `
[
  <button
    class=""
    data-testid="sort-button"
  >
    популярности
  </button>,
  <button
    class="active"
    data-testid="sort-button"
  >
    цене
  </button>,
  <button
    class=""
    data-testid="sort-button"
  >
    алфавиту
  </button>,
]
`;

